# Project 1

The first thing we need to do in order to analyze this data is read in the csv file and store it as a pandas dataframe. 

In [109]:
import numpy as np
import pandas as pd

df = pd.read_csv('ontime_delayedflights.csv')

df

            city  airline  ontime_count  delayed_count
0    Los Angeles   Alaska           497             62
1    Los Angeles  AM West           694            117
2        Phoenix   Alaska           221             12
3        Phoenix  AM West          4840            415
4      San Diego   Alaska           212             20
5      San Diego  AM West           383             65
6  San Francisco   Alaska           503            102
7  San Francisco  AM West           320            129
8        Seattle   Alaska          1841            305
9        Seattle  AM West           201             61

We are going to compare the average number of on time and delayed flights for each airline. In order to do that we need to create two new data frames which contain the separate flight data for each airline. We will do that by using boolean logic to caputre only the rows where the airline is equal to the value we want.  

In [110]:
al = df[df.airline == 'Alaska']
am = df[df.airline == 'AM West']

Now we can calculate the averages

For Alaska:

In [112]:
al[['ontime_count', 'delayed_count']].mean()

ontime_count     654.8
delayed_count    100.2
dtype: float64

For AM West:

In [113]:
am[['ontime_count', 'delayed_count']].mean()

ontime_count     1287.6
delayed_count     157.4
dtype: float64

As you can see Alaska Airlines had an average of 654.8 ontime flights and 100.2 delayed flights over the five cities. 

AM West had an average of 1287.6 ontime flights and 157.4 delayed. 

Although AM west had a higher rate of ontime flights they also had a higher rate of delayed flights. This is probably due to conducting a higher number of flights over all. 

pandas can help us to verify this. We will use the cumsum() method to create a series which adds each subsequent column value to the next. We will the use idxmax to give us the index location of the total sum which we will then add together. 

In [114]:
alontime = al.ontime_count.cumsum()
aldelayed = al.delayed_count.cumsum()

amontime = am.ontime_count.cumsum()
amdelayed = am.delayed_count.cumsum()

alotmax = alontime.idxmax()
aldmax = aldelayed.idxmax()
amotmax = amontime.idxmax()
amdmax = amdelayed.idxmax()

In [115]:
altotal = alontime.loc[alotmax] + aldelayed.loc[aldmax]
amtotal = amontime.loc[amotmax] + amontime.loc[amdmax]

In [116]:
altotal

3775

In [117]:
amtotal

12876

Now we can see that AM West flew almost triple the amount of total flights of Alaska in our data. This is probably what accounts for the much higher rate of ontime flights. 

There are other comparisons we can do as well. 

Let's find out which city and airline had the most flights over all. .idxmax() will give us the index location of both the max ontime and delayed flights.

In [118]:
df[['ontime_count', 'delayed_count']].idxmax()

ontime_count     3
delayed_count    3
dtype: int64

Now we can look up this index to see the whole picture. 

In [119]:
df.loc[3]

city             Phoenix
airline          AM West
ontime_count        4840
delayed_count        415
Name: 3, dtype: object

AM West flights out of Phoenix were responsible for the highest number of ontime and delayed flights in this dataset. 

Now we can do the reverse and find the index of the minimum values.

In [120]:
df[['ontime_count', 'delayed_count']].idxmin()

ontime_count     9
delayed_count    2
dtype: int64

For the minimum values we get two different index locations.

In [121]:
df.loc[9]

city             Seattle
airline          AM West
ontime_count         201
delayed_count         61
Name: 9, dtype: object

In [122]:
df.loc[2]

city             Phoenix
airline           Alaska
ontime_count         221
delayed_count         12
Name: 2, dtype: object

So the minimum amount of ontime flights were from AM West in Seattle and the minimum delayed flights were from Alaska Airlines in Phoenix. 

We can do the same operations to find the Min and Max values individually using our separated dataframes. 

In [123]:
al[['ontime_count', 'delayed_count']].idxmax()

ontime_count     8
delayed_count    8
dtype: int64

In [124]:
al.loc[8]

city             Seattle
airline           Alaska
ontime_count        1841
delayed_count        305
Name: 8, dtype: object

In [125]:
al[['ontime_count', 'delayed_count']].idxmin()

ontime_count     4
delayed_count    2
dtype: int64

In [126]:
al.loc[4]

city             San Diego
airline             Alaska
ontime_count           212
delayed_count           20
Name: 4, dtype: object

In [127]:
al.loc[2]

city             Phoenix
airline           Alaska
ontime_count         221
delayed_count         12
Name: 2, dtype: object

In [128]:
am[['ontime_count', 'delayed_count']].idxmax()

ontime_count     3
delayed_count    3
dtype: int64

In [129]:
am.loc[3]

city             Phoenix
airline          AM West
ontime_count        4840
delayed_count        415
Name: 3, dtype: object

In [130]:
am[['ontime_count', 'delayed_count']].idxmin()

ontime_count     9
delayed_count    9
dtype: int64

In [131]:
am.loc[9]

city             Seattle
airline          AM West
ontime_count         201
delayed_count         61
Name: 9, dtype: object

This analysis leads us to conclude that these airlines operate most frequently in the regions they are named for. Alaska Airlines most flights operate out of Seatlle and it's least operate out of San Diego and Phoenix. AM West operates mostly out of Phoenix and minimally out of Seattle. 

We can also see that although AM West has a higher rate of on time flights this is due to the large number of flights they operate out of Phoenix. We can also see that the rate of delayed flights is quite similar for both airlines. 